enhancements/ignition-spec-dual-support.md

```diff
@@ -0,0 +1,447 @@
+update the ignition version on RHCOS/Installer/MCO to make use of the changes.
+
+This proposal will also allow closer alignment with OKD, as OKD will be based
+on Fedora CoreOS (Henceforth FCoS), which is already on, and only supports,
+ignition spec 3. We want to do this in a way that can minimize deltas between
+OKD and OCP.
+
+
+### Goals
+
+#### Phase 1:
+- [ ] A config translator is created to translate from ignition spec 2 to spec 3
+- [ ] A config translator is created to translate from ignition spec 3 to spec 2
+- [ ] The MCO is updated to support both ignition spec 2 and 3, with:
+    - [ ] MCC rendering spec 3 configs to spec 2
+    - [ ] MCD gaining the ability to process both spec 2 and spec 3 configs
+- [ ] Create tests to apply spec 3 configs to spec 2 clusters successfully
+- [ ] An alerting mechanism is put in place for outdated and incompatible/non-translatable configs
+- [ ] Docs are updated to reflect the new config version
+
+#### Phase 1 (optional):
+- [ ] MCC attempts to translate all machineconfigs with v2 ignition config to v3, leaving those which cannot be converted and warning the user of future changes
+- [ ] MCC learns to generate v3 initial master/worker ignition configs
+
+#### Phase 2:
+- [ ] The MCO gains the ability to manage installer-generated stub master/worker ignition configs (separate enhancement proposal)
+- [ ] ignition-dracut spec 2 and spec 3 diffs are aligned
+- [ ] RHCOS bootimages switches to only accept ignition spec 3 configs
+- [ ] The Openshift installer is updated to generate spec 3 configs
+- [ ] The MCC gains the ability for users to provide necessary changes to update spec 2 to spec 3
+- [ ] MCO enforces that all configs are spec 3 before allowing the CVO to start the update
+- [ ] Further tests/docs are added
+
+
+### Non-Goals
+
+ - Support FCoS/OKD directly through this change
+ - API support for MCO, namely switching to RawExt formatted machineconfig
+ objects instead of explicitly referencing ignition, is not considered as
+ part of this proposal
+
+
+## Proposal
+
+This change is multi-component:
+
+
+### Vendoring Changes
+
+The MCO and Installer must change to go modules (currently dep) for vendoring
+as ignition v2 requires using go modules. To support both spec 2 and spec 3,
+both ignition versions must be vendored in for typing. 
+
+
+### Spec Translator
+
+To handle both spec versions, as we need the ability to upgrade existing
+clusters, we will create a translator between spec 2 and spec 3. This ensures
+that a cluster only has one “desiredConfig” which will be translated to spec 3
+when the MCO with dual support detects that the existing configuration of a
+machine is on spec 2 (will happen only once for all existing and new nodes,
+when the MCO with dual support is first deployed onto the cluster). This will
+only be required as part of the MCO.
+
+Note that there exists three types of spec 2 configs:
```

Comment:
For the purpose of this document, there are two types: those that we are certain we can translate and everything else. We aren’t going to handle the uncertain and certain-failure cases differently (are we?).

Reply:
Updated